### PR TITLE
Fix moving of files

### PR DESCRIPTION
**Pull request by Fifi-wang**
We fix the issue that the program does not move files according to the user's settings.
Refs [#7679](https://github.com/JabRef/jabref/issues/7679)

- [X] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [X] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [X] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java [hunk 1]
```diff
@@ -59,6 +59,10 @@ public boolean moveToDefaultDirectory() throws IOException {
                     filePreferences.getFileDirectoryPattern());
         }
 
+        if (filePreferences.getFileDirectory().isPresent()) {
+            targetDirectoryName = filePreferences.getFileDirectory().get().toString() + "\\" + targetDirectoryName;
```
Comment (koppor) [inline src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java:63]: `\\` is a Windows path separator.

**Review by koppor — CHANGES_REQUESTED** [review]
I do not think that his fixes the issue. See the comment at https://github.com/JabRef/jabref/issues/7679#issuecomment-830068789 that is a **Mac OS X** issue. See https://bugs.openjdk.java.net/browse/JDK-8264172. Could you solve that?

Moreover, there need to be **test cases**.

**Comment by Fifi-wang** [comment]
> I do not think that his fixes the issue. See the comment at [#7679 (comment)](https://github.com/JabRef/jabref/issues/7679#issuecomment-830068789) that is a **Mac OS X** issue. See https://bugs.openjdk.java.net/browse/JDK-8264172. Could you solve that?
> 
> Moreover, there need to be **test cases**.

I'm sorry that I didn't notice this detail, because I mainly tested on windows computer. In fact, this problem also exists on windows computer, that is, the saved path does not conform to the setting of main file directory.

File: src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java (hunk 1 above)
Comment (Fifi-wang) [reply src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java]: Thank you for raising this critical question. This time I used File.separator to ensure the separator of different systems.

File: src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java [hunk 2]
```diff
@@ -59,6 +60,10 @@ public boolean moveToDefaultDirectory() throws IOException {
                     filePreferences.getFileDirectoryPattern());
         }
 
+        if (filePreferences.getFileDirectory().isPresent()) {
+            targetDirectoryName = filePreferences.getFileDirectory().get().toString() + File.separator + targetDirectoryName;
```
Comment (koppor) [inline src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java:64]: Please use Java 8 API to construct 

```suggestion
            targetDirectoryName = filePreferences.getFileDirectory().get().resolve(targetDirectoryName).toString();
```

**Comment by koppor** [comment]
@Fifi-wang Thank you for the follow up. Please add test cases. With different filePreferences and (possibly) [Junit TempDir](https://www.baeldung.com/junit-5-temporary-directory). The reason for test cases is that we want to ensure that behavior remains controlled for the code. If someone touches that code again, the tests should be green or modified on purpose. Currently, there seem no tests for `LinkedFileHandler`. Press <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>T</kbd> on IntelliJ to create a test. This is a a good exercise IMHO to get known to unit testing.

File: src/test/java/org/jabref/logic/externalfiles/LinkedFileHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    private LinkedFileHandler linkedFileHandler;
+    private FilePreferences filePreferences;
+
+    @BeforeEach
+    void setUp(@TempDir Path testFolder) throws IOException {
+        Path bibPath = testFolder.resolve("bib_test.bib");
+        Path pdfPath = testFolder.resolve("pdf_test.pdf");
+        Files.createFile(bibPath);
+        Files.createFile(pdfPath);
+        try (OutputStreamWriter outputStreamWriter = new OutputStreamWriter(new FileOutputStream(bibPath.toAbsolutePath().toString()), StandardCharsets.UTF_8)) {
```
Comment (koppor) [inline src/test/java/org/jabref/logic/externalfiles/LinkedFileHandlerTest.java:36]: Please use `Files.writeString` (see https://howtodoinjava.com/java11/write-string-to-file/)

File: src/test/java/org/jabref/logic/externalfiles/LinkedFileHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        assert file.exists();
+    }
+
+    @Test
+    void moveToDefaultDirectoryWithoutMainFileDirectory(@TempDir Path testFolder) throws IOException {
+        when(filePreferences.getFileDirectoryPattern()).thenReturn("");
+        when(filePreferences.getFileDirectory()).thenReturn(Optional.empty());
+        linkedFileHandler.moveToDefaultDirectory();
+        File file = testFolder.resolve("pdf_test.pdf").toFile();
+        assert file.exists();
```
Comment (koppor) [inline src/test/java/org/jabref/logic/externalfiles/LinkedFileHandlerTest.java:72]:
```suggestion
        assertTrue(Files.exists(testFolder.resolve("pdf_test.pdf")))
```

See https://stackoverflow.com/a/5130959/873282

File: src/test/java/org/jabref/logic/externalfiles/LinkedFileHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        );
+    }
+
+    @Test
+    void moveToDefaultDirectoryWithMainFileDirectory(@TempDir Path testFolder) throws IOException {
+        when(filePreferences.getFileDirectoryPattern()).thenReturn("papers");
+        when(filePreferences.getFileDirectory()).thenReturn(Optional.of(testFolder.resolve("main")));
+        linkedFileHandler.moveToDefaultDirectory();
+        File file = testFolder.resolve("main/papers/pdf_test.pdf").toFile();
+        assert file.exists();
```
Comment (koppor) [inline src/test/java/org/jabref/logic/externalfiles/LinkedFileHandlerTest.java:63]:
```suggestion
        assertTrue(Files.exists(estFolder.resolve("main/papers/pdf_test.pdf")))
```

See https://stackoverflow.com/a/5130959/873282

File: src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java [hunk 6]
```diff
@@ -38,7 +38,7 @@ public LinkedFileHandler(LinkedFile fileEntry,
         this.databaseContext = Objects.requireNonNull(databaseContext);
         this.filePreferences = Objects.requireNonNull(filePreferences);
     }
-
+
```
Comment (koppor) [inline src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java:41]: If possible, remove the spaces - they are not needed here. Maybe, <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>L</kbd> helps to reformat (and delete thus spaces) easily.

**Review by koppor — CHANGES_REQUESTED** [review]
Thank you so much for adding tests.

Note that `assertTrue` is used in Junit (https://www.petrikainulainen.net/programming/testing/junit-5-tutorial-writing-assertions-with-junit-5-api/)

The `assert` statements are build-in java commands only available when debugging is active.

File: src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java (hunk 6 above)
Comment (Fifi-wang) [reply src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java]: I agree with you. That's my mistake.

File: src/test/java/org/jabref/logic/externalfiles/LinkedFileHandlerTest.java (hunk 3 above)
Comment (Fifi-wang) [reply src/test/java/org/jabref/logic/externalfiles/LinkedFileHandlerTest.java]: OK, I will have try.

**Comment by Fifi-wang** [comment]
@koppor I'm sorry that I have no idea that why I can't pass the unit tests because I didn't change any thing in LinkFileHandler.java. Could you give me some suggestions?

**Comment by calixtus** [comment]
The currently failing tests are the HelpFileTests, they are not connected to your changes. Maybe try to merge the latest main branch into your branch.

File: src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java [hunk 7]
```diff
@@ -59,6 +59,10 @@ public boolean moveToDefaultDirectory() throws IOException {
                     filePreferences.getFileDirectoryPattern());
         }
 
+        if (filePreferences.getFileDirectory().isPresent()) {
+            targetDirectoryName = filePreferences.getFileDirectory().get().resolve(targetDirectoryName).toString();
+        }
```
Comment (k3KAW8Pnf7mkmdSMPHz27) [inline src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java:64]: Isn't this duplicating the behavior of line 43?

```java
Optional<Path> targetDirectory = databaseContext.getFirstExistingFileDir(filePreferences);
```

If the `FileDirectory` should be used, isn't it already in the content of `targetDirectory`?

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Probably refs. #7658

**Comment by Fifi-wang** [comment]
> The currently failing tests are the HelpFileTests, they are not connected to your changes. Maybe try to merge the latest main branch into your branch.

Thanks a lot! I will try it later!

File: src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java (hunk 7 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java:64]: Yep, the addition does not make any sense

Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/externalfiles/LinkedFileHandler.java:64]: And even worse, it overwrites the behavior of the  setting "should store files relative to libary file location"
 overwrites all other directories, so the other tests are actually useless

**Review by Siedlerchr — CHANGES_REQUESTED** [review]
Sorry for the late reply, but this is not a solution. You are actually changing intended behaviour.

**Pull request closed without merging**